File: src/main/java/org/example/simpletest/other/rudong/DeviceParser.java
package org.example.simpletest.other.rudong;

import com.google.common.collect.Lists;
import jxl.Cell;
import jxl.Sheet;
import jxl.Workbook;
import jxl.write.Label;
import jxl.write.WritableSheet;
import jxl.write.WritableWorkbook;
import jxl.write.WriteException;
import lombok.Builder;
import lombok.Data;
import org.apache.commons.lang3.StringUtils;
import org.springframework.core.io.ClassPathResource;

import java.io.File;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

/**
 * @author zhuzhenjie
 * @since 2021/1/18
 */
public class DeviceParser {


    static List<String> xlsNames = Lists.newArrayList(
            "如东_栟茶", "如东_丰利",
            "如东_河口", "如东_洋口", "如东_袁庄");

    public static void main(String[] args) {
//        xlsNames.stream().map(DeviceParser::onePlantDevice).forEach(DeviceParser::writeToNew);
        xlsNames.forEach(xlsName -> {
            writeToNew(onePlantDevice(xlsName), xlsName);
        });
    }

    @Builder
    @Data
    static class DeviceTmp {
        private String group;
        private String name;
    }

    static void writeToNew(Map<String, List<DeviceTmp>> deviceInfos, String plantName) {
        if (deviceInfos == null || deviceInfos.size() <= 0) {
            return;
        }
        String tmpPath = "D:/tmp/project_rudong/" + plantName + "_modify.xls";

        try {
            File file = new File(tmpPath);
            file.createNewFile();
            WritableWorkbook wb = Workbook.createWorkbook(file);
            // 创建sheet
            WritableSheet sheet = wb.createSheet("sheet", 0);

            int groupCol = 0;
            int nameCol = 1;
            int monitorItemCnCol = 2;

            Label group = new Label(groupCol, 0, "设备名称");
            Label monitorItemAll = new Label(nameCol, 0, "监测项全名");
            Label monitorItem = new Label(monitorItemCnCol, 0, "监测项中文");

            sheet.addCell(group);
            sheet.addCell(monitorItemAll);
            sheet.addCell(monitorItem);


            AtomicInteger begin = new AtomicInteger(1);
            deviceInfos.values().stream().flatMap(Collection::stream).forEach(
                    deviceTmp -> {
                        String deviceGroup = deviceTmp.getGroup();
                        String deviceName = deviceTmp.getName();
                        int row = begin.getAndIncrement();
                        Label tmpGroup = new Label(groupCol, row, deviceGroup);
                        Label tmpName = new Label(nameCol, row, deviceName);
                        try {
                            sheet.addCell(tmpGroup);
                            sheet.addCell(tmpName);

                            // 监测项的配置
                            String monitorItemCn = deviceName.contains(deviceGroup) ? deviceName.replace(deviceGroup, StringUtils.EMPTY) : StringUtils.EMPTY;
                            sheet.addCell(new Label(monitorItemCnCol, row, monitorItemCn));
                        } catch (WriteException e) {
                            e.printStackTrace();
                        }
                    }
            );

            wb.write();
            wb.close();
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    static Map<String, List<DeviceTmp>> onePlantDevice(String xlsName) {
        try {
            String classpath = "project_rudong/xls/" + xlsName + ".xls";
            Workbook workbook = Workbook.getWorkbook(new ClassPathResource(classpath).getInputStream());
            Sheet sheet = workbook.getSheet(0);
            int rows = sheet.getRows();
            List<DeviceTmp> all = Lists.newArrayList();
            // 第一行是表头
            for (int i = 1; i < rows; i++) {
                Cell[] row = sheet.getRow(i);
                // 第一列是name 第二列是group
                all.add(DeviceTmp.builder()
                        .group(row[1].getContents().trim())
                        .name(row[0].getContents().trim())
                        .build());
            }
            return all.stream().collect(Collectors.groupingBy(DeviceTmp::getGroup));
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: src/main/kotlin/org/example/simpletest/kotlin/other/rudong/DeviceParser.kt
@file:Suppress("NAME_SHADOWING")

package org.example.simpletest.kotlin.other.rudong

import jxl.Cell
import jxl.Sheet
import jxl.Workbook
import jxl.write.Label
import jxl.write.WritableSheet
import jxl.write.WritableWorkbook
import org.apache.commons.lang3.StringUtils
import org.example.simpletest.kotlin.getLogger
import org.springframework.core.io.ClassPathResource
import java.io.File

/**
 * 如东设备解析
 *
 * @author zhuzhenjie
 * @since 2021/1/18
 */
private val log = getLogger("rudong-device-parses")


private val xlsNames = listOf(
    "如东_栟茶", "如东_丰利",
    "如东_河口", "如东_洋口", "如东_袁庄"
)

private const val tmpFolder = "D:/tmp/project_rudong/"


fun main() {
//    init()
//    xlsNames.forEach { createTransXls(createBasicXls(onePlantDevice(it), it), it) }

    xlsNames.forEach { translate(getTranslateMap(it), it) }
}


fun translate(trans: Map<String, String>, xlsName: String) {
    if (trans.isEmpty()) return
    val path = "${tmpFolder}${xlsName}_basic.xls"
    try {
        val file = File(path)
        val wb = Workbook.getWorkbook(file)
        val wwb = Workbook.createWorkbook(file, wb)

        val wSheet = wwb.getSheet(0)

        val cnColNum = 2
        val enColNum = 3
        IntRange(1, wSheet.rows - 1).forEach {
            val cell = wSheet.getCell(cnColNum, it)
            val cn = cell.contents.trim()

            val en = if (StringUtils.isNotBlank(cn) && trans.containsKey(cn)) trans[cn] else StringUtils.EMPTY

            if (StringUtils.isNotBlank(en)) {
                wSheet.addCell(Label(enColNum, it, en))
            }

        }
        wwb.write()
        wwb.close()
        wb.close()
    } catch (e: Exception) {
        log.error(StringUtils.EMPTY, e)
    }
}

fun getTranslateMap(xlsName: String): Map<String, String> {
    val classpath = "project_rudong/xls/translate/${xlsName}_translate.xls"
    val wb = Workbook.getWorkbook(ClassPathResource(classpath).inputStream)
    val st = wb.getSheet(0)
    val map: MutableMap<String, String> = mutableMapOf<String, String>()
    IntRange(0, st.rows - 1).forEach { map[st.getCell(0, it).contents.trim()] = st.getCell(1, it).contents.trim() }
    return map
}

/**
 *
 * 新建需要翻译的英文部分
 */
fun createTransXls(cnSet: Set<String>?, plantName: String) {
    if (null == cnSet || cnSet.isEmpty()) {
        return
    }
    val folder = File("${tmpFolder}translate")
    if (!folder.exists()) {
        folder.mkdirs()
    }
    val tmpPath = "${tmpFolder}translate/${plantName}_translate.xls"
    try {
        val file = File(tmpPath)
        file.createNewFile()
        val wb = Workbook.createWorkbook(file)
        val sheet = wb.createSheet("sheet1", 0)

        val cnCol = 0
        var rowBegin = 0
        cnSet.forEach { sheet.addCell(Label(cnCol, rowBegin++, it)) }

        wb.write()
        wb.close()
    } catch (e: Exception) {
        e.printStackTrace()
    }

}

/**
 *
 */
fun createBasicXls(deviceInfos: Map<String, List<DeviceTmp>>?, plantName: String): Set<String> {
    if (deviceInfos == null || deviceInfos.isEmpty()) {
        return emptySet()
    }
    val tmpPath = "${tmpFolder}${plantName}_basic.xls"
    try {
        val file = File(tmpPath)
        file.createNewFile()
        val wb: WritableWorkbook = Workbook.createWorkbook(file)
        // 创建sheet
        val sheet: WritableSheet = wb.createSheet("sheet1", 0)

        val groupCol = 0
        val nameCol = 1
        val monitorItemCnCol = 2

        // 设置title
        sheet.addCell(Label(groupCol, 0, "设备名称"))
        sheet.addCell(Label(nameCol, 0, "监测项全名"))
        sheet.addCell(Label(monitorItemCnCol, 0, "监测项中文名"))
        sheet.addCell(Label(3, 0, "监测项英文名"))

        var begin = 1

        val cnSet = mutableSetOf<String>()

        deviceInfos.forEach { (_, deviceTmpList) ->
            deviceTmpList.forEach {
                val deviceGroup = it.group
                val deviceName = it.name
                val insertRowNum = begin++
                try {
                    sheet.addCell(Label(groupCol, insertRowNum, deviceGroup))
                    sheet.addCell(Label(nameCol, insertRowNum, deviceName))

                    // 监测项目中文名
                    fun String.addInSet(): String {
                        cnSet.add(this)
                        return this
                    }

                    val monitorItemCn = if (deviceName.contains(deviceGroup) && deviceName.length > deviceGroup.length)
                        deviceName.replace(deviceGroup, StringUtils.EMPTY).addInSet() else StringUtils.EMPTY

                    sheet.addCell(Label(monitorItemCnCol, insertRowNum, monitorItemCn))
                } catch (e: Exception) {
                    log.error(StringUtils.EMPTY, e)
                }
            }
        }
        wb.write()
        wb.close()
        return cnSet.filter { StringUtils.isNotBlank(it) }.toHashSet()
    } catch (e: Exception) {
        e.printStackTrace()
        return emptySet()
    }
}


fun onePlantDevice(xlsName: String): MutableMap<String, MutableList<DeviceTmp>>? {
    try {
        val classPath = "project_rudong/xls/${xlsName}.xls"
        val wb: Workbook = Workbook.getWorkbook(ClassPathResource(classPath).inputStream)
        val sheet: Sheet = wb.getSheet(0)
        val rows = sheet.rows

        val all: MutableList<DeviceTmp> = mutableListOf()

        // 第一行是表头
        for (i in 1 until rows) {
            val rows: Array<Cell> = sheet.getRow(i)
            all.add(DeviceTmp(rows[1].contents.trim(), rows[0].contents.trim()))
        }
        // kotlin 写法
        return all.groupByTo(HashMap(), { it.group })
    } catch (e: Exception) {
        e.printStackTrace()
        return null
    }
}

fun init() {
    val basicFolder = File(tmpFolder)
    if (!basicFolder.exists()) {
        basicFolder.mkdirs()
    }
}

data class DeviceTmp(val group: String, val name: String)
